### PR TITLE
[ML][Inference] Adding inference ingest processor

### PR DESCRIPTION
**Pull request by benwtrent**
This adds a new Ingest processor that does infers against a stored previously trained model.

### CODE REVIEW

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/MachineLearning.java [hunk 1]
```diff
@@ -416,7 +429,9 @@ public static boolean isMlNode(DiscoveryNode node) {
                 AutodetectBuilder.MAX_ANOMALY_RECORDS_SETTING_DYNAMIC,
                 MAX_OPEN_JOBS_PER_NODE,
                 MIN_DISK_SPACE_OFF_HEAP,
-                MlConfigMigrationEligibilityCheck.ENABLE_CONFIG_MIGRATION);
+                MlConfigMigrationEligibilityCheck.ENABLE_CONFIG_MIGRATION,
+                InferenceProcessor.MAX_INFERENCE_PROCESSORS
```
Comment (benwtrent) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/MachineLearning.java]: There may be one more setting we should add "Maximum loaded models". But, I think that we don't need to add this other setting until we support loading models outside of processors.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/ingest/InferenceProcessor.java [hunk 2]
```diff
@@ -5,28 +5,126 @@
...
+
+import static org.elasticsearch.xpack.core.ClientHelper.ML_ORIGIN;
+import static org.elasticsearch.xpack.core.ClientHelper.executeAsyncWithOrigin;
+
 
 public class InferenceProcessor extends AbstractProcessor {
 
+    // How many total inference processors are allowed to be used in the cluster.
+    public static final Setting<Integer> MAX_INFERENCE_PROCESSORS = Setting.intSetting("xpack.ml.max_inference_processors",
+        50,
```
Comment (benwtrent) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/ingest/InferenceProcessor.java]: this is a "magic" number. Given we don't have real data around our typical model size and performance, I just picked a number that is not too small.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/ingest/InferenceProcessor.java [hunk 3]
```diff
@@ -38,4 +136,120 @@ public IngestDocument execute(IngestDocument ingestDocument) {
...
+                    currentInferenceProcessors,
+                    MAX_INFERENCE_PROCESSORS.getKey(),
+                    maxIngestProcessors);
+            }
+
+            String modelId = ConfigurationUtils.readStringProperty(TYPE, tag, config, MODEL_ID);
+            String targetField = ConfigurationUtils.readStringProperty(TYPE, tag, config, TARGET_FIELD);
+            Map<String, String> fieldMapping = ConfigurationUtils.readOptionalMap(TYPE, tag, config, FIELD_MAPPINGS);
+            InferenceConfig inferenceConfig = inferenceConfigFromMap(ConfigurationUtils.readMap(TYPE, tag, config, INFERENCE_CONFIG));
+            String modelInfoField = ConfigurationUtils.readStringProperty(TYPE, tag, config, MODEL_INFO_FIELD, "_model_info");
```
Comment (benwtrent) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/ingest/InferenceProcessor.java]: By default, I think we should add a field that includes the model ID used in the inference step. 

This should probably append the `tag` to the end of the default info field to protect against multiple processors in the same pipeline...

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/ingest/InferenceProcessor.java [hunk 4]
```diff
@@ -38,4 +136,120 @@ public IngestDocument execute(IngestDocument ingestDocument) {
...
+                checkSupportedVersion(new RegressionConfig());
+                return RegressionConfig.fromMap(valueMap);
+            } else {
+                throw ExceptionsHelper.badRequestException("unrecognized inference configuration type {}. Supported types {}",
+                    inferenceConfig.keySet(),
+                    Arrays.asList(ClassificationConfig.NAME, RegressionConfig.NAME));
+            }
+        }
+
+        void checkSupportedVersion(InferenceConfig config) {
```
Comment (benwtrent) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/ingest/InferenceProcessor.java]: Initially, this factory builds the pipeline on the master node before storing it in cluster state. So, if we always check the minimum supported version against the minimum node version we can guarantee that `PUT pipeline` will fail if a user tries to create a pipeline when the specific processor setting is not supported.

File: x-pack/plugin/ml/qa/native-multi-node-tests/src/test/java/org/elasticsearch/xpack/ml/integration/InferenceIngestIT.java [hunk 5]
```diff
@@ -0,0 +1,530 @@
...
+
+    public void testPipelineCreationAndDeletion() throws Exception {
+
+        for (int i = 0; i < 10; i++) {
+            assertThat(client().admin().cluster().preparePutPipeline("simple_classification_pipeline",
+                new BytesArray(CLASSIFICATION_PIPELINE.getBytes(StandardCharsets.UTF_8)),
+                XContentType.JSON).get().isAcknowledged(), is(true));
+
+            client().prepareIndex("index_for_inference_test", "_doc")
+                .setSource(new HashMap<>(){{
```
Comment (przemekwitek) [inline x-pack/plugin/ml/qa/native-multi-node-tests/src/test/java/org/elasticsearch/xpack/ml/integration/InferenceIngestIT.java]: I would put source doc generation into a method, say, "generateDocSource".

File: x-pack/plugin/ml/qa/native-multi-node-tests/src/test/java/org/elasticsearch/xpack/ml/integration/InferenceIngestIT.java [hunk 6]
```diff
@@ -0,0 +1,530 @@
...
+                    .size(0)
+                    .trackTotalHits(true)
+                    .query(QueryBuilders.boolQuery()
+                        .filter(QueryBuilders.existsQuery("result_class"))))).get().getHits().getTotalHits().value,
+            equalTo(20L));
+
+    }
+
+    public void testSimulate() {
+        String source = "{\n" +
```
Comment (przemekwitek) [inline x-pack/plugin/ml/qa/native-multi-node-tests/src/test/java/org/elasticsearch/xpack/ml/integration/InferenceIngestIT.java]: [non-actionable] Waiting for multiline raw string literals being introduced to Java...

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/ingest/InferenceProcessor.java [hunk 7]
```diff
@@ -53,26 +140,122 @@ public String getType() {
...
+        // Package private for testing
+        void setMaxIngestProcessors(int maxIngestProcessors) {
+            logger.trace("updating setting maxIngestProcessors from [{}] to [{}]", this.maxIngestProcessors, maxIngestProcessors);
+            this.maxIngestProcessors = maxIngestProcessors;
+        }
+
+        InferenceConfig inferenceConfigFromMap(Map<String, Object> inferenceConfig) throws IOException {
+            ExceptionsHelper.requireNonNull(inferenceConfig, INFERENCE_CONFIG);
+
+            if (inferenceConfig.keySet().size() != 1) {
```
Comment (przemekwitek) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/ingest/InferenceProcessor.java]: Is it equivalent to `inferenceConfig.size()`? It should be, right?

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/ingest/InferenceProcessor.java [hunk 8]
```diff
@@ -53,26 +140,122 @@ public String getType() {
...
+            Object value = inferenceConfig.values().iterator().next();
+
+            if ((value instanceof Map<?, ?>) == false) {
+                throw ExceptionsHelper.badRequestException("{} must be an object with one inference type mapped to an object.",
+                    INFERENCE_CONFIG);
+            }
+            @SuppressWarnings("unchecked")
+            Map<String, Object> valueMap = (Map<String, Object>)value;
+
+            if (inferenceConfig.containsKey(ClassificationConfig.NAME)) {
```
Comment (przemekwitek) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/ingest/InferenceProcessor.java]: I'm wondering how many places in the code will need to be changed when if we introduce a third inference type. Maybe not a problem now, but something to keep in mind if we decide the type based on a string here.

File: x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/inference/ingest/InferenceProcessorFactoryTests.java [hunk 9]
```diff
@@ -0,0 +1,266 @@
...
+        }};
+
+        try {
+            processorFactory.create(Collections.emptyMap(), "my_inference_processor", classification);
+        } catch (Exception ex) {
+            fail(ex.getMessage());
+        }
+    }
+
+    private static ClusterState buildState(MetaData metaData) {
```
Comment (przemekwitek) [inline x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/inference/ingest/InferenceProcessorFactoryTests.java]: Rename to `buildClusterState` for consistency with the methods below?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/RegressionConfig.java [hunk 10]
```diff
@@ -5,16 +5,27 @@
...
 import java.util.Objects;
 
 public class RegressionConfig implements InferenceConfig {
 
     public static final String NAME = "regression";
+    private static final Version MIN_SUPPORTED_VERSION = Version.V_8_0_0;
+
+    public static RegressionConfig fromMap(Map<String, Object> map) {
+        if (map.isEmpty() == false) {
+            throw ExceptionsHelper.badRequestException("Unrecognized fields {}.", map.keySet());
```
[the remark this answers is not in the document]
Comment (benwtrent) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/RegressionConfig.java]: @przemekwitek no. a `keySet().toString()` writes like this `[key1, key2,...]`

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/ingest/InferenceProcessor.java [hunk 11]
```diff
@@ -5,28 +5,126 @@
...
+                if (srcValue != null) {
+                    fields.put(dest, srcValue);
+                }
+            });
+        }
+        return new InferModelAction.Request(modelId, fields, inferenceConfig);
+    }
+
+    void mutateDocument(InferModelAction.Response response, IngestDocument ingestDocument) {
+        response.getInferenceResults().get(0).writeResult(ingestDocument, this.targetField);
```
[the remark this answers is not in the document]
Comment (benwtrent) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/ingest/InferenceProcessor.java]: Good catch, I will put a check to protect us from funkiness.

File: x-pack/plugin/ml/qa/native-multi-node-tests/src/test/java/org/elasticsearch/xpack/ml/integration/InferenceIngestIT.java (hunk 6 above)
Comment (benwtrent) [reply x-pack/plugin/ml/qa/native-multi-node-tests/src/test/java/org/elasticsearch/xpack/ml/integration/InferenceIngestIT.java]: No joke!
```
"""
"""
```
cannot be added soon enough!

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/ingest/InferenceProcessor.java (hunk 7 above)
Comment (benwtrent) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/ingest/InferenceProcessor.java]: definitely, I can simplify.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/ingest/InferenceProcessor.java (hunk 8 above)
Comment (benwtrent) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/ingest/InferenceProcessor.java]: what sucks is that the inference processor objects are not `toXContent` objects. So, I cannot use the existing named objects.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/InferenceConfig.java [hunk 12]
```diff
@@ -13,4 +14,5 @@
 
     boolean isTargetTypeSupported(TargetType targetType);
 
+    Version getMinimalSupportedVersion();
```
Comment (przemekwitek) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/InferenceConfig.java]: Could you add a comment explaining the need for this?

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/ingest/InferenceProcessor.java (hunk 4 above)
Comment (przemekwitek) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/ingest/InferenceProcessor.java]: I think this explanation should go into code comment for method `getMinimalSupportedVersion`.

**Review by przemekwitek — APPROVED** [review]
LGTM

**Pull request merged**